src/material/core/datetime/date-selection-model.ts

```diff
@@ -0,0 +1,207 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {FactoryProvider, Injectable, Optional, SkipSelf, OnDestroy} from '@angular/core';
+import {DateAdapter} from './date-adapter';
+import {Observable, Subject} from 'rxjs';
+
+/** A class representing a range of dates. */
+export class DateRange<D> {
+  constructor(
+    /** The start date of the range. */
+    readonly start: D | null,
+    /** The end date of the range. */
+    readonly end: D | null) {}
+}
+
+type ExtractDateTypeFromSelection<T> = T extends DateRange<infer D> ? D : NonNullable<T>;
+
+/** Event emitted by the date selection model when its selection changes. */
+export interface DateSelectionModelChange<S> {
+  /** New value for the selection. */
+  selection: S;
+
+  /** Object that triggered the change. */
+  source: any;
+}
+
+/** A selection model containing a date selection. */
+export abstract class MatDateSelectionModel<S, D = ExtractDateTypeFromSelection<S>>
+    implements OnDestroy {
+  private _selectionChanged = new Subject<DateSelectionModelChange<S>>();
+
+  /** Emits when the selection has changed. */
+  selectionChanged: Observable<DateSelectionModelChange<S>> =
+  this._selectionChanged.asObservable();
+
+  /** The current selection. */
+  readonly selection: S;
+
+  protected constructor(protected readonly adapter: DateAdapter<D>, selection: S) {
+    this.selection = selection;
+  }
+
+  updateSelection(value: S, source: any) {
+    (this as {selection: S}).selection = value;
+    this._selectionChanged.next({selection: value, source});
+  }
+
+  ngOnDestroy() {
+    this._selectionChanged.complete();
+  }
+
+  /** Adds a date to the current selection. */
+  abstract add(date: D | null): void;
+
+  /** Checks whether the current selection is complete. */
+  abstract isComplete(): boolean;
+
+  /**
+   * Checks whether the current selection is the same as the selection in the given selection model.
```

Comment:
Not sure if this was a change you made or not, but the comment doesn't seem to line up with the types. The parameter is a selection not a selection model